### ISSUE TITLE
azurerm_container_group always detects changes when deploying to virtual networks

### ISSUE THREAD

**Issue by rossbeehler** [hashicorp/terraform-provider-azurerm#11544]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.15.1
on windows_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.56.0

### Affected Resource(s)
azurerm_container_group 

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = ">= 2.56"
    }
  }
  backend "azurerm" {
    resource_group_name  = <redacted>
    storage_account_name = <redacted>
    container_name       = <redacted>
    key                  = <redacted>
  }
}

provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "rg" {
  name     = <redacted>
  location = "eastus2"
}

data "azurerm_subnet" "sn" {
  name                 = <redacted>
  virtual_network_name = <redacted>
  resource_group_name  = <redacted>
}

resource "azurerm_network_profile" "np" {
  name                = "netprofile"
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name

  container_network_interface {
    name = "cnic"

    ip_configuration {
      name      = "ipconfig"
      subnet_id = data.azurerm_subnet.sn.id
    }
  }
}

resource "azurerm_container_group" "cg" {
  name                = "containers"
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name
  ip_address_type     = "Private"
  network_profile_id  = azurerm_network_profile.np.id
  os_type             = "Linux"

  container {
    name   = "container1"
    image  = <redacted>
    cpu    = "0.5"
    memory = "1.5"
    environment_variables = <redacted>
  }

  container {
    name   = "container2"
    image  = <redacted>
    cpu    = "0.5"
    memory = "1.5"
    ports {
      port     = 443
      protocol = "TCP"
    }
    volume {
      name       = <redacted>
      mount_path = <redacted>
      read_only  = true
      secret = <redacted>
    }
  }

  image_registry_credential {
    server   = <redacted>
    username = <redacted>
    password = <redacted>
  }

  restart_policy = "OnFailure"

  dns_config {
    nameservers    = [<redacted>]
    search_domains = [<redacted>]
    options        = []
  }

  tags = {
    environment = "tag1"
  }
}
```

### Expected Behaviour

When running `terraform apply` twice consecutively, no changes should be detected on the second run.

### Actual Behaviour

It finds changes like the following, noting I've removed some extra noise:

``` 
  # azurerm_container_group.cg must be replaced
-/+ resource "azurerm_container_group" "cg" {
      - dns_name_label      = "" -> null

      ~ container {
          - secure_environment_variables = (sensitive value)
            # (4 unchanged attributes hidden)
        }
      ~ container {
          - environment_variables        = {} -> null
          - secure_environment_variables = (sensitive value)
          ~ volume {
                name                 = "secrets"
              - share_name           = "" -> null
              - storage_account_key  = (sensitive value)
              - storage_account_name = "" -> null
                # (4 unchanged attributes hidden)
            }
        }

      ~ dns_config {
          - options        = [
              - "",
            ] -> null # forces replacement
        }

      + identity {
          + identity_ids = (known after apply)
          + principal_id = (known after apply)
          + type         = (known after apply)
        }

        # (1 unchanged block hidden)
    }
```

### Steps to Reproduce

1. `terraform apply`
2. `terraform apply`

### Important Factoids

Tried adding "empty" properties like `dns_name_label`, but that isn't supported when connecting to a virtual network.

Some of the other identified changes may have nothing to do with the virtual network.

**Comment by favoretti** [comment]
Hi there and thank you for reporting this. Can you try setting `dns_config`'s `options` value to `null`, rather than an empty array (`[]`)? This should abolish the diff.

**Comment by rossbeehler** [comment]
Hi @favoretti ,

When I set `options` to `null`, I get this error:

```
╷
│ Error: "dns_config.0.options": required field is not set
│
│   with azurerm_container_group.cg,
│   on container_group.tf line 1, in resource "azurerm_container_group" "cg":
│    1: resource "azurerm_container_group" "cg" {
│
╵
```

That said, the provider should be able to equate an empty array to null even if that error wasn't received, and all the other differences for `dns_label`, `container.environment_variables`, `container.secure_environment_variables`, `container.volume.share_name`, `container.volume.storage_account_name`, and `container.volume.storage_account_key` should equate `""` with `null`.  Do note that I tried also setting each of those properties explicitly to `null`, and they still showed up as differences like this: `- dns_name_label      = "" -> null`

**Comment by favoretti** [comment]
I see, thank you for elaborating. I'll try to run some tests to see how to fix this.

**Comment by rossbeehler** [comment]
Hi @favoretti, any updates on this?  Note that this prevents anyone from quickly rotating secret environment variable/volume values without a complete re-deploy, which is a very slow operation.

**Comment by sinc59** [comment]
Hello @favoretti did you have any news about this issue ?
I would like update a terraform registry module with the dns_config argument.

**Comment by stephybun** [comment]
Closed by #15618
